Sprint 12

kazakovdmitriy opened the pull request.

Commit:
Добавлен ImageListService

Commit:
Fix fetchPhotosNextPage

Commit:
fix

Commit:
Add ProfileLogoutService

Commit:
Добавил анимацию на загрузку изображений из сети

Helvetii951 commented on ImageFeed/Helpers/UIView+Extensions.swift:
```diff
@@ -20,4 +20,33 @@ extension UIView {
+    
+    func removeAnimation() {
+            layer.sublayers?.removeAll { $0.name == "animationGradient" }
+        }
```
Чуть поехали отступы. Они помогают коду выглядеть аккуратнее и привычнее взгляду, так читать код намного удобнее. Чтобы поправить оступы: можно выделить весь код `cmd+A` и зажать `contol+i`.

Helvetii951 commented on ImageFeed/ImagesList/ImagesListCell.swift:
```diff
@@ -7,27 +7,57 @@
+    // MARK: - IB Actions
+    @IBAction func didTapLikeButton(_ sender: Any) {
+        delegate?.imageListCellDidTapLike(self)
+    }
```
Функцию можно сделать приватной. Рекомендую всегда писать функции и свойства с минимальной областью видимости, а затем менять ее, если нужно.

Helvetii951 commented on ImageFeed/Profile/ProfileViewController.swift:
```diff
@@ -137,7 +148,37 @@ final class ProfileViewController: UIViewController {
+        let splashViewController = SplashViewController()
+        splashViewController.modalPresentationStyle = .fullScreen
+        present(splashViewController, animated: true)
+    }
```
Здесь стоит устанавливать`rootViewController`. Пусть старые контроллеры выметутся из памяти.

Helvetii951 commented on ImageFeed/ImagesList/ImagesListViewController.swift:
```diff
@@ -21,49 +27,119 @@ final class ImagesListViewController: UIViewController {
-                let viewController = segue.destination as? SingleImageViewController, // 2
-                let indexPath = sender as? IndexPath // 3
+                let viewController = segue.destination as? SingleImageViewController,
+                let indexPath = sender as? IndexPath
             else {
-                assertionFailure("Invalid segue destination") // 4
+                assertionFailure("Invalid segue destination")
                 return
             }
-
-            let image = UIImage(named: photosName[indexPath.row]) // 5
-            viewController.image = image // 6
+            
+            viewController.imageURLString = photos[indexPath.row].largeImageURL
         } else {
             super.prepare(for: segue, sender: sender) // 7
         }
     }
-    
+
+    // MARK: - Public Methods
     func configCell(for cell: ImagesListCell, with indexPath: IndexPath) {
+        
+        cell.delegate = self
+        
+        let photo = photos[indexPath.row]
+        
+        guard let url = URL(string: photo.thumbImageURL) else { return }
+        
```
`guard`ом можно отсечь дальнейшую настройку ячейки таблицы. Я бы рекомендовал if-let.

Helvetii951 commented on ImageFeed/ImagesList/ImagesListViewController.swift:
```diff
@@ -21,49 +27,119 @@ final class ImagesListViewController: UIViewController {
+        let photo = photos[indexPath.row]
+        
+        UIBlockingProgressHUD.show()
+        imageListService.changeLike(photoId: photo.id, isLike: !photo.isLiked) { [weak self] result in
+            
+            guard let self = self else { return }
+            
+            switch result {
+            case .success:
+                self.photos = self.imageListService.photos
+                cell.setIsLiked(isLiked: self.photos[indexPath.row].isLiked)
+                UIBlockingProgressHUD.dismiss()
+            case .failure(let error):
+                print("Ошибка изменения лайка: \(error.localizedDescription)")
+                UIBlockingProgressHUD.dismiss()
+                self.showError()
+            }
+        }
+        cell.setIsLiked(isLiked: photo.isLiked)
```
`cell.setIsLiked(...)` вызовется до выполнения запроса. Так мы получим ошибочное состояние. Запрос, не прошел, но лайк мы визуально проставили. Может запутать пользователя.

Commit:
Убрал анимацию градиента
избавился от лишнего dateFormatter
Сделал выход из функции, если уже идет закачка списка фото
SplashViewController при логауте устанавливается как rootViewController

Commit:
Добавил аллерт с ошибкой в splahViewController
